### PR TITLE
Initial sqlotus implementation (schema, populator, recency-tracker, logs )

### PR DESCRIPTION
**Pull request by ribasushi**
=== Update 2021-06-14, detailing status of the core of the sqlotus/instalotus/sql-baed-visor code

( disclaimer based on prior disagreements with the sentinel team: the pieces missing described below are hard-requirements from my PoV. I will be extremely unhappy/be compelled to "do it right on my own time" if someone picks this up to ship a reduced-scope version )

This code in this PR represents a PoC of a large-scale blockstore, that is in turn consumed by 3 commits here: https://github.com/ribasushi/ltsh/commits/sqlotus

That in turn was supposed to enable a PL-hosted distributed Pg-replication service, which would be consumed by anumber of collabs and also feed things like https://github.com/protocol/web3-dev-team/pull/29

This PR is incomplete/remains marked as draft due to a few architectural problems together with performance issues, which were acutely exposed during the cron-overload period. Specifically:

- The Put() implementation is done in mutex-guarded series, and is currently too slow ( nearing 10s per batch on the high end of ~20k blocks per batch ) under very large spikes of activity. There is a way to prallelize the puts into multiple independent index pages, while keeping a single lock for the duration of a Put() ( locks are required due to a quirk of Pg's MVCC: https://stackoverflow.com/a/42217872 )

- The ancestry record system ( the linkparser ) needs to be reevaluated in light of the above

- The orphan-tracking recursive function needs to be rewritten to utilize indices better - it is currently a slight bottleneck at over 1s per headchange

- The library needs to be split into 2 parts, one plain blockstore, one filecoin chainstore "subclass", as the general blockstore-with-ancestors is usable externally, allowing for extremely efficient write-time-refcounting and thus very efficient GC.

=== End update

This is the backend implements of a so called 'sqlotus'. This blockstore
implements the classic IPFS blockstore interface with a number of extras:

- Transparent intelligent compresssion of data where it offers benefits
- Every block is parsed and its intermediate links are recorded, forming in-db DAG
- For every encountered-but-not-yet-available CID a "mapping stub" record is kept
- The blockstore takes a tipset as a context and keeps detailed block r/w logs
- On every context change it logs a minimal functional chainstate overview

The specific methods of interest are:

- An accessor for the underlying PostgreSQL connection for ad-hoc queries
```
  func (dbbs *PgBlockstore) PgxPool() *pgxpool.Pool
```

- Several flavors of Get()-ers for single, bulk and background retrieval
```
  func (dbbs *PgBlockstore) GetStoredBlock(c cid.Cid) (*StoredBlock, error)
  func (dbbs *PgBlockstore) InflateDbRows(rows pgx.Rows, eagerUnpack bool, skipCaching bool) (blocks []*StoredBlock, bytesCached int64, err error)
  func (dbbs *PgBlockstore) PrecacheByOrdinal(ctx context.Context, ordinals []int64) (totalCount, totalBytes int64, err error)
```

- A method to timestamp all to-date block-accesses with an epoch and tipset
```
  func (dbbs *PgBlockstore) FlushAccessLogs(contextEpoch abi.ChainEpoch, contextTipsetDbOrdinal *int32) error
```

- Filecoin chain-specific bookkeeping
```
  func (dbbs *PgBlockstore) GetFilTipSetHead(ctx context.Context) (*DestructuredFilTipSetData, error)
  func (dbbs *PgBlockstore) FindFilTipSet(ctx context.Context, tipsetCids []cid.Cid, walkBackEpochs abi.ChainEpoch) (*DestructuredFilTipSetData, error)
  func (dbbs *PgBlockstore) StoreFilTipSetData(ctx context.Context, tsd *DestructuredFilTipSetData) (tipsetDbOrdinal *int32, err error)
  func (dbbs *PgBlockstore) StoreFilTipSetVisit(ctx context.Context, tsDbOrdinal *int32, tsEpoch abi.ChainEpoch, visitAt time.Time, isHeadChange bool) (err error)
```

Motivations for this design are:

- Simplicity: stock postgres permits very high velocity/low overhead replication,
  No extra extensions or applications required, permitting robust full-archival

- Auditability: the extra logs kept as the chainsync machinery progresses allows
  examining the exact sequence of events when the network is in distress.

- Performance: while the blockstore performs badly in generic benchmarks, it
  outperforms a traditional store by strategic prefetching of recently-accessed
  metadata on startup. Additionally the DAG-records allow for fine-grained
  prefetch during runtime operation: e.g. chan exports and/or GraphQL backends

- Debuggability: the "mapping stubs" and access logs permit analysis of the
  otherwise vast amount of data. Ideally when a full chain is available, this
  will allow for more robust abalysis of access hotspots, in turn enabling more
  informed IPLD optimizations

### CODE REVIEW

File: track_fil_state.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+	TypedSignature           []byte
+	ForkSignalVarint         []byte
+	TicketProof              []byte
+	ElectionWincountAndProof []byte
+	WinpostTypesAndProof     [][]byte
+}
+
+// CurrentFilTipSetKeyBytes queries the configured instance namespace and
+// retrieves the epoch and ordered CIDs of the most-recently Visit()-ed tipset.
+func (dbbs *PgBlockstore) CurrentFilTipSetKey(ctx context.Context) ([]cid.Cid, abi.ChainEpoch, error) {
```
Comment (willscott) [inline track_fil_state.go]: might be nice to also expose the 'get tipset n behind current tipset' (the lookback view used by e.g. visor)

Comment (ribasushi) [reply track_fil_state.go]: @willscott: how is this?

```
// GetFilTipsetCurrent queries the configured instance namespace and retrieves
// the epoch and ordered CIDs of the most-recently Visit()-ed tipset.
func (dbbs *PgBlockstore) GetFilTipsetCurrent(ctx context.Context) ([]cid.Cid, abi.ChainEpoch, error) { ... }

// FindFilTipset queries the shared tipset list for the tipset formed from the
// provided chain header block CIDs. If the optional walkBackEpochs is larger
// than 0, return the parent at that distance from the initial given tipset. 
func (dbbs *PgBlockstore) FindFilTipset(ctx context.Context, tipsetCids []cid.Cid, walkBackEpochs abi.ChainEpoch) ([]cid.Cid, abi.ChainEpoch, error) { ... }
```

Comment (willscott) [reply track_fil_state.go]: that would be great

Comment (ribasushi) [reply track_fil_state.go]: @willscot switched this up a bit more: https://github.com/filecoin-project/go-bs-postgres-chainnotated/commit/cc21260af03ef4bfed6ba4b0c8a61fed8871a19f#diff-a7784a9261f9c1e3044236d29c9919f789463b57051b784afd90e02c7283bc92R42-R100

TLDR:
https://pkg.go.dev/github.com/filecoin-project/go-bs-postgres-chainnotated@v0.0.0-20210307170649-cc21260af03e#PgBlockstore.FindFilTipSet
+
https://pkg.go.dev/github.com/filecoin-project/go-bs-postgres-chainnotated@v0.0.0-20210307170649-cc21260af03e#PgBlockstore.GetFilTipSetHead

**Comment by ribasushi** [comment]
@iand I know it's a tall order ( this is a large codebase ) but if time permits I'd love to have your thoughts. This is a preliminary version, a number of things aren't as optimal as they can be, yada yada. If something looks odd it almost certainly is, please point it out :)

**Comment by iand** [comment]
> @iand I know it's a tall order ( this is a large codebase ) but if time permits I'd love to have your thoughts. This is a preliminary version, a number of things aren't as optimal as they can be, yada yada. If something looks odd it almost certainly is, please point it out :)

Would be a pleasure!

File: globals.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	"regexp"
+	"runtime"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+)
+
+// Filecoin Mainnet constants redefined here to avoid import cycles.
+const (
```
Comment (iand) [inline globals.go:13]: Does this prevent use of the blockstore for calibration and similar networks?

File: globals.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	MaxPgxPoolSize                  = int32(32)             // the maximum size of the connection pool (default scales dynamically based on NumCPU)
+	TrackRecentTipsetsCount         = 25 * EpochsInHour     // determines how many Filecoin epochs in the past do we consider a block to be "recently-accessed"
+	MaxFutureEpochsAsRecent         = 3 * 24 * EpochsInHour // When backtracking over long distances, prune "recent" entries that end up being that far into the future
+	MaxOrphanLookback               = 1 + ChainFinality     // when keeping track of orphaned states, how far back do we want to walk
+	AssumeUserNeverMutatesBlockdata = true                  // for performance mimic the behavior of github.com/ipfs/go-block-format regarding data ownership and copying
+)
+
+const (
+	DefaultLruCacheSize           = int64(32 << 30)                                                      // the default LRU cache size in bytes, overridable in PgBlockstoreConfig
+	BulkFetchSize                 = 256 << 10                                                            // how many blocks to fetch at once from various RDBMS sets
```
Comment (iand) [inline globals.go]: `32 << 30` would convey meaning better with a constant for gigabytes. Not sure what the significance of `<< 10` is for a count of blocks. Maybe just write that as `1 << 18` or is there some significance to 256 that I am missing?

File: lib/cidkeyedlru/cklru.go [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+		return false
+	}
+
+	if cost <= 0 {
+		panic("Add() with negative or zero cost unsupported")
+	}
+
+	l.lookupsMu.Lock()
+	l.orderedMu.Lock()
+	defer func() { l.lookupsMu.Unlock(); l.orderedMu.Unlock() }()
```
Comment (iand) [inline lib/cidkeyedlru/cklru.go]: These unlocks appear to be in the wrong order

File: lib/cidkeyedlru/cklru.go [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+		l.metrics.CountEvicted++
+	}
+
+	newEntry := &LruEntry{
+		Cost:   cost,
+		Cid:    c,
+		Object: object,
+	}
+
+	if len(subDict) == 0 {
```
Comment (iand) [inline lib/cidkeyedlru/cklru.go:180]: You could take this simple path for small subdicts,  e.g. < 100

File: lib/cidkeyedlru/cklru.go [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+			continue
+		}
+
+		if toInsertGroups[tb] == nil {
+			toInsertGroups[tb] = make([]*LruEntry, 0, 4096)
+		}
+		toInsertGroups[tb] = append(toInsertGroups[tb], entries[i])
+
+		if entries[i].Cost <= 0 {
+			panic("Populate() with negative or zero cost unsupported")
```
Comment (iand) [inline lib/cidkeyedlru/cklru.go:229]: This panic will leave lookupsMu locked

File: lib/cidkeyedlru/cklru.go [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+
+	l.lookupsMu.RUnlock()
+
+	if newCost == 0 {
+		return
+	}
+
+	l.lookupsMu.Lock()
+	l.orderedMu.Lock()
+	defer func() { l.lookupsMu.Unlock(); l.orderedMu.Unlock() }()
```
Comment (iand) [inline lib/cidkeyedlru/cklru.go]: Wrong unlock order?

File: lib/cidkeyedlru/cklru.go [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+
+			// perform the check once more with all locks now in place
+			// contention here can at times be very severe
+			if _, found := subDict[entry.Cid]; found {
+				continue
+			}
+
+			// don't allow an overflow
+			if spaceLeft-entry.Cost < 0 {
+				return
```
Comment (iand) [inline lib/cidkeyedlru/cklru.go]: Returning here will prevent a lower cost entry being added. If the high cost entry was first then the whole batch is ignored.

File: lib/cidkeyedlru/cklru.go [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+				}
+				subDict[entry.Cid] = l.ordered.InsertBefore(entry, randomElt) // InsertBefore() == closer-to-Front()
+			}
+
+			l.metrics.CostAdded += entry.Cost
+			l.metrics.CountAdded++
+		}
+	}
+
+	return
```
Comment (iand) [inline lib/cidkeyedlru/cklru.go]: Not needed :)

File: lib/cidkeyedlru/cklru.go [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+	if !found {
+		return false
+	}
+
+	delete(l.lookups[tailByte(c)], c)
+
+	l.metrics.CostEvicted += l.ordered.Remove(elt).(*LruEntry).Cost
+	l.metrics.CountEvicted++
+
+	l.lookupsMu.Unlock()
```
Comment (iand) [inline lib/cidkeyedlru/cklru.go]: Unlock in reverse. My reasoning is that although the code now may be consistent it makes it easier for future changes to accidentally deadlock. For example, perhaps someone inserts a function call after lookups is unlocked that tries to take the locks in the usual order: now it is deadlocked on orderedMu.

File: lib/synccid/synccidset.go [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+// Set provides a kludge-y concurrency-safe version of cid.Set
+//
+// Just like the original it is an implementation of a set of Cids,
+// which holds a single copy of every Cids that is added to it.
+//
+// Unlike the original there is no constructor: the zero value of a
+// Set struct is ready to be used as a new empty Set
```
Comment (iand) [inline lib/synccid/synccidset.go:18]: :heart:

File: lib/synccid/synccidset.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+// the result was a new addition or false if the set
+// already contained the given Cid
+func (cs *Set) Visit(c cid.Cid) (isNewMember bool) {
+	// do not record the empty cid
+	if c.KeyString() == "" {
+		return false
+	}
+
+	cs.mu.Lock()
+
```
Comment (iand) [inline lib/synccid/synccidset.go:65]: Would just be cleaner to use defer unlock here and similar methods on this type

File: lib/synccid/synccidset.go [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+	if _, found := cs.set[c.KeyString()]; !found {
+		cs.set[c.KeyString()] = struct{}{}
+		isNewMember = true
+	}
+
+	cs.mu.Unlock()
+	return
+}
+
+// Remove deletes a Cid from the Set
```
Comment (iand) [inline lib/synccid/synccidset.go]: comment: one or several Cids

File: lib/synccid/synccidset.go [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+
+	cs.mu.Unlock()
+}
+
+// Len returns how many elements the Set has
+func (cs *Set) Len() int {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if cs.set == nil {
```
Comment (iand) [inline lib/synccid/synccidset.go]: len of a nil map is zero

File: store.go [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+	}
+}
+
+//
+// Start of blockstore.Blockstore implementation
+//
+
+// DeleteBlock is NOT IMPLEMENTED by this store - the storage is strictly
+// append-only. Always returns an error upon invocation.
+func (NIM *PgBlockstore) DeleteBlock(cid.Cid) (err error) {
```
Comment (iand) [inline store.go:121]: nit: use consistent receiver name

File: store_io.go [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+		sb := &StoredBlock{
+			cid:          rootCid,
+			size:         int32(len(dmh.Digest)),
+			blockContent: dmh.Digest,
+		}
+
+		return sb, nil
+	}
+
+	// relatively hot path, save an alloc
```
Comment (iand) [inline store_io.go:99]: Seems odd that you care about this allocation but then use Sprintf to build the query each time

File: store.go [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+		return cb(data)
+	}
+}
+
+// AllKeysChan returns a channel which will in turn provide every CID currently
+// in the RDBMS. Do not use this method under normal circumstances - every call
+// would create an MVCC snapshot on the database-side and potentially consume
+// enormous amount of resources, until either the channel is exhausted, or the
+// initially provided context is cancelled.
+func (dbbs *PgBlockstore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
```
Comment (iand) [inline store.go:230]: Would love to be able to deprecate this function in the underlying interface. Good way to tie up resources by calling and then just partially reading from the returned channel.

File: store_bulk_ops.go [hunk 18]
```diff
@@ -0,0 +1,686 @@
...
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/dustin/go-humanize"
+	"github.com/jackc/pgx/v4"
+	"golang.org/x/xerrors"
+)
+
+// Background initial state prefetch
+// Panic()s on early errors
```
Comment (iand) [inline store_bulk_ops.go:24]: Why panic rather than return an error? Even though it is run in a background goroutine it could return an error to that goroutine which could decide to log and ignore or panic itself

File: store_bulk_ops.go [hunk 19]
```diff
@@ -0,0 +1,686 @@
...
+					rows.Close()
+				}
+				errs <- err
+				return
+			}
+
+			sbs, bytes, err := dbbs.InflateDbRows(rows, false, false)
+			if err != nil {
+				errs <- err
+				return
```
Comment (iand) [inline store_bulk_ops.go:193]: Does rows.Close need to be called here as it is above?

File: store_init.go [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+			dbSettings, dbSettingsExpected,
+		)
+	}
+
+	// check whether we are talking to a slave, and force RO to shut off tipset-tracking automatically
+	if dbbs.isWritable {
+		if _, err := dbPool.Exec(ctx, fmt.Sprintf(
+			"CREATE TEMPORARY TABLE %s ( pk INTEGER ) ON COMMIT DROP",
+			"tmptable_"+randBytesAsHex(),
+		)); err != nil {
```
Comment (iand) [inline store_init.go]: Check type of error here to avoid transient errors forcing RO mode.

File: store_init.go [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+
+		if err = dbbs.deploy(ctx, neededDDL); err != nil {
+			return nil, err
+		}
+	}
+
+	// not RO and everything is deployed: let's make sure we can:
+	// - EX-lock the namespace or quit
+	// - then auto-SH-lock it going forward
+	if dbbs.isWritable && dbbs.instanceNamespace != "" {
```
Comment (iand) [inline store_init.go:209]: I don't understand this handover sequence. You attempt to take an exclusive lock then wait on a shared lock in a new connection until the exclusive lock connection has dropped, then retry the exclusive lock. What is the purpose of the second try?

File: store_init.go [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+			)
+		}
+
+		if !cfg.AutoUpdateSchema {
+			return nil, xerrors.New(
+				"unable to continue: currently-deployed schema does not (partially) match the expected state, and AutoUpdateSchema is not enabled",
+			)
+		}
+
+		if err = dbbs.deploy(ctx, neededDDL); err != nil {
```
Comment (iand) [inline store_init.go:201]: This deploy happens before you have gained exclusive access to the database?

File: store_init.go [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+
+	if cfg.ExtraPreloadNamespace != "" {
+		if !validNamespace.MatchString(cfg.ExtraPreloadNamespace) {
+			return nil, fmt.Errorf("provided instance-specific extra namespace '%s' does not match %s", cfg.ExtraPreloadNamespace, validNamespace.String())
+		}
+		dbbs.additionalPreloadNamespace = "fil_instance_" + cfg.ExtraPreloadNamespace
+	}
+
+	if cfg.CacheSizeGiB != 0 {
+		if cfg.CacheSizeGiB > 256 {
```
Comment (iand) [inline store_init.go:56]: Why is 256GiB too large?

File: store_init.go [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+// in line with the provided PgBlockstoreConfig. See the documentation of this
+// struct for more info.
+func NewPgBlockstore(ctx context.Context, cfg PgBlockstoreConfig) (chainAnnotatedPgBlockstore *PgBlockstore, err error) {
+
+	shutdownChan := make(chan struct{})
+	dbbs := &PgBlockstore{
+		cacheInactiveBeforeRead:  cfg.CacheInactiveBeforeRead,
+		isWritable:               cfg.StoreIsWritable,
+		parseBlockLinks:          !cfg.DisableBlocklinkParsing,
+		lruSizeBytes:             DefaultLruCacheSize,
```
Comment (iand) [inline store_init.go]: Consider making the default a % of available memory?

File: store_io.go [hunk 25]
```diff
@@ -0,0 +1,600 @@
...
+			// https://github.com/jackc/pgx/issues/845#issuecomment-705550012
+			sb.dbDeflatedContent = append(append(
+				make([]byte, 0, len(zstdMagic)+len(dbContentBytes)),
+				zstdMagic...,
+			), dbContentBytes...)
+
+			// we know we will need the content soon: trigger decompression in the background
+			// do so without a limiter: if we starve the CPU for a bit - so be it
+			if eagerUnpack {
+				go sb.inflatedContent(false)
```
Comment (iand) [inline store_io.go]: This seems wasteful considering the attention to small optimizations elsewhere. What if the fetch of all rows fails just before the last one? There's no way to stop these goroutines. Also the number of goroutines is unbounded with unknown memory allocation for each.

**Review by iand — COMMENTED** [review]
Reviewed for as long as I could but didn't make it all the way through :)

File: store_io.go (hunk 25 above)
Comment (ribasushi) [reply store_io.go]: You are correct, remnant from an ancient iteration.
Removed from function call and placed at the right spot here: https://github.com/filecoin-project/go-bs-postgres-chainnotated/commit/8b1f2e9296c4346ca7647889cde5d2ecae123f8e#diff-97588be80ee1efc9827cef3de764f82d5ede8bdc7ead07d610268e49f84e70d6L129-R182

File: store_init.go (hunk 22 above)
Comment (ribasushi) [reply store_init.go:201]: > exclusive access to the database?

There is no exclusive lock on the db, only on namespaces ( there can and often are multiple writers on the same RDBMS, each with its own namespace )

The entire deploy is guarded by a serializable transaction here: https://github.com/filecoin-project/go-bs-postgres-chainnotated/blob/8b1f2e9296c4346ca7647889cde5d2ecae123f8e/store_schema.go#L569-L592
It will either happen or not at all

Let me know if this doesn't answer the concern in full.

File: store_init.go (hunk 21 above)
Comment (ribasushi) [reply store_init.go:209]: @iand the thinking is that this snippet:
https://github.com/filecoin-project/go-bs-postgres-chainnotated/blob/8b1f2e9296c4346ca7647889cde5d2ecae123f8e/store_init.go#L211-L220
could get tripped up by a *different* piece of code ( perhaps not even written by me ) grabbing a shared lock as soon as the other exlock quits.
Basically I am making sure that the first connection from the new pool can properly upgrade itself, indicating that yes indeed the entire dance worked. Perhaps I am being overly paranoid and should just remove the entire "double-check" section.

Thoughts?

File: lib/cidkeyedlru/cklru.go (hunk 5 above)
Comment (ribasushi) [reply lib/cidkeyedlru/cklru.go:180]: This is not as much a "simpler path" but rather a `if ThingWeGoBefore == nil ...`
Given the cache is virtually never at such a low level ( it is designed, and s quickly filled with >million blocks), I am inclined to keep it as-is.

File: store_init.go (hunk 21 above)
Comment (ribasushi) [reply store_init.go:209]: @iand I am a muppet, after sleeping on it - you are 100% correct
Fixed in https://github.com/filecoin-project/go-bs-postgres-chainnotated/pull/1/commits/0edaa329d98

File: lib/cidkeyedlru/cklru.go (hunk 6 above)
Comment (ribasushi) [reply lib/cidkeyedlru/cklru.go:229]: Sigh you are right
`panic` != `exit` :(

File: lib/cidkeyedlru/cklru.go (hunk 10 above)
Comment (ribasushi) [reply lib/cidkeyedlru/cklru.go]: Fair point, adjusted all sites to unlock in reverse.

File: lib/synccid/synccidset.go (hunk 12 above)
Comment (ribasushi) [reply lib/synccid/synccidset.go:65]: I *really* dislike using defer in short methods when I can avoid it.

I can be persuaded otherwise if you have a decent argument though.

File: lib/cidkeyedlru/cklru.go (hunk 8 above)
Comment (ribasushi) [reply lib/cidkeyedlru/cklru.go]: Good point, switched to `continue`

File: lib/synccid/synccidset.go (hunk 13 above)
Comment (ribasushi) [reply lib/synccid/synccidset.go]: Docs adjusted

File: lib/synccid/synccidset.go (hunk 14 above)
Comment (ribasushi) [reply lib/synccid/synccidset.go]: Huh... indeed! I was expecting an NPE

File: lib/synccid/synccidset.go (hunk 12 above)
Comment (iand) [reply lib/synccid/synccidset.go:65]: Intent is clearer and it's idiomatic for those who come after. There is a still a tiny performance hit for using defer so it's not always appropriate.

File: store.go [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+	}
+}
+
+//
+// Start of blockstore.Blockstore implementation
+//
+
+// DeleteBlock is NOT IMPLEMENTED by this store - the storage is strictly
+// append-only. Always returns an error upon invocation.
+func (NIM *PgBlockstore) DeleteBlock(cid.Cid) (err error) {
```
Comment (frrist) [inline store.go:121]: Driveby comment: with recent changes in lotus (if this is trying to match the lotus blockstore interface(?)) this will need to be extended to meet the new interface: https://github.com/filecoin-project/lotus/blob/4c05ec28ba45e6947118c5a0bc836f312442d0e9/blockstore/blockstore.go#L20-L21

Comment (ribasushi) [reply store.go:121]: Yeah definitely,haven't had a chance to upgrade past 1.5.0 yet
Keep the comments coming - I am slowly working through them ( this PR will be split into 3 as a result of some other discussions, but all the same: I do read )

File: store_init.go (hunk 20 above)
Comment (ribasushi) [reply store_init.go]: Good catch, fixed in https://github.com/filecoin-project/go-bs-postgres-chainnotated/pull/1/commits/7ebc314e42a

File: store.go (hunk 26 above)
Comment (ribasushi) [reply store.go:121]: Addressed in https://github.com/filecoin-project/go-bs-postgres-chainnotated/pull/1/commits/6f5650b22ffa01ff1d2c021916d85d4e7e91e072#diff-ed648181f98484bd12541509e0ae7b5ad1d1de7674ab1721814b47ddd5c95de4R124-R131

File: store_bulk_ops.go (hunk 19 above)
Comment (ribasushi) [reply store_bulk_ops.go:193]: @iand no because the inflator does that. I renamed the method to make it more sensible: https://github.com/filecoin-project/go-bs-postgres-chainnotated/pull/1/commits/84c363ecaa6
